### PR TITLE
Delete event_push_summary_unique_index again.

### PR DESCRIPTION
**Pull request by clokep**
See https://github.com/matrix-org/synapse/issues/14641#issuecomment-1344243412 for details, but if a Synapse install was bumped such that certain schema deltas all ran at the same time we could have:

1. Registered a background update to add an index.
2. Deleted the index if it exists (but it won't since the background job hasn't run).
3. Assumed the index was deleted.

We have two fixes for this which target different situations:

1. Drop the background job from 1 above -- it is no longer needed and will stop newly updated Synapses from breaking.
2. Fix any broken Synapses by dropping the obsolete index again.

### CODE REVIEW

**Review by DMRobertson — COMMENTED** [review]
An unfortunate footgun.

> if a Synapse install was bumped such that certain schema deltas all ran at the same time we could have:
> 
>     1. Registered a background update to add an index.
> 
>     2. Deleted the index if it exists (but it won't since the background job hasn't run).
> 
>     3. Assumed the index was deleted.

with hindsight, could we have some something different here to avoid this? I guess we should have deleted the background job from (1) in migration (2)... but I'm not convinced that's bulletproof. E.g. what if that background job is running while we try to delete it?

**Review by DMRobertson — APPROVED** [review]

**Comment by clokep** [comment]
@DMRobertson Actually, now that I think about this. Should we remove the code which adds a background index for the old update?

**Comment by DMRobertson** [comment]
> Should we remove the code which adds a background index for the old update?

Err, I'm having some difficulty following all the details. I think you mean this?

https://github.com/matrix-org/synapse/blob/e8bce8999f21d30affc459755e304a1f4732165c/synapse/storage/databases/main/event_push_actions.py#L277-L284

I think(?) it makes sense to remove this in version X if
- there is never a situation where we'd want to run this as part of upgrading to X, and
- no other background updates block waiting for this one to finish.

**Comment by clokep** [comment]
> Err, I'm having some difficulty following all the details. I think you mean this?
> 
> https://github.com/matrix-org/synapse/blob/e8bce8999f21d30affc459755e304a1f4732165c/synapse/storage/databases/main/event_push_actions.py#L277-L284

Yeah, pretty much. It seems that replaces another index though, so I'm not 100% sure what to do about that. (And actually, deleting the background job could theoretically leave a *different* old index I think?)

**Comment by DMRobertson** [comment]
Ahh, I think I'm starting to understand.

I think this drop

https://github.com/matrix-org/synapse/blob/dbf18f514ea5d2539ba3148049eae5a6793f1d60/synapse/storage/schema/main/delta/73/06thread_notifications_thread_id_idx.sql#L16-L17

isn't safe to do in a migration like this. It's dropping something that was created in a background update---to robustly do so we need to do the drop in a background update that depends on the first.

**Comment by clokep** [comment]
> Ahh, I think I'm starting to understand.
> 
> I think this drop
> 
> https://github.com/matrix-org/synapse/blob/dbf18f514ea5d2539ba3148049eae5a6793f1d60/synapse/storage/schema/main/delta/73/06thread_notifications_thread_id_idx.sql#L16-L17
> 
> isn't safe to do in a migration like this. It's dropping something that was created in a background update---to robustly do so we need to do the drop in a background update that depends on the first.

Yeah, I'm starting to think that's the right thing to do. Will update this PR.

**Comment by clokep** [comment]
Alright, I made a couple of tweaks which I think are safe:

* Also delete the index which `event_push_summary_unique_index` replaces.
* Delete the code which registers the background job since it is unused.

File: synapse/storage/schema/main/delta/73/23_fix_thread_index.sql [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+-- to be created in the background from delta 71/02event_push_summary_unique.sql after
+-- delta 73/06thread_notifications_thread_id_idx.sql is executed, causing it to
+-- not drop the event_push_summary_unique_index index.
+--
+-- See https://github.com/matrix-org/synapse/issues/14641
+
+-- Stop the index from being scheduled for creation in the background.
+DELETE FROM background_updates WHERE update_name = 'event_push_summary_unique_index';
+
+-- The above background job also replaces another index, if blah ensure that side-effect
```
Comment (asymmetric) [inline synapse/storage/schema/main/delta/73/23_fix_thread_index.sql:27]: Is `blah` there on purpose?

File: synapse/storage/databases/main/event_push_actions.py [hunk 2]
```diff
@@ -274,15 +274,6 @@ def __init__(
...
 
-        self.db_pool.updates.register_background_index_update(
-            "event_push_summary_unique_index",
-            index_name="event_push_summary_unique_index",
-            table="event_push_summary",
-            columns=["user_id", "room_id"],
-            unique=True,
-            replaces_index="event_push_summary_user_rm",
-        )
-
```
Comment (DMRobertson) [inline synapse/storage/databases/main/event_push_actions.py:285]: Sanity check: what does the background update logic do if it tries to run an update but there's no handler for it?

Oh, the migration deletes the update in question, so it doesn't matter. :+1:

File: synapse/storage/schema/main/delta/73/23_fix_thread_index.sql (hunk 1 above)
Comment (DMRobertson) [reply synapse/storage/schema/main/delta/73/23_fix_thread_index.sql:27]:
```suggestion
-- The above background job also replaces another index, so ensure that side-effect
```

File: synapse/storage/schema/main/delta/73/23_fix_thread_index.sql [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+-- If a Synapse deployment made a large jump in versions (from < 1.62.0 to >= 1.70.0)
+-- in a single upgrade then it might be possible for the event_push_summary_unique_index
+-- to be created in the background from delta 71/02event_push_summary_unique.sql after
+-- delta 73/06thread_notifications_thread_id_idx.sql is executed, causing it to
+-- not drop the event_push_summary_unique_index index.
+--
+-- See https://github.com/matrix-org/synapse/issues/14641
+
+-- Stop the index from being scheduled for creation in the background.
+DELETE FROM background_updates WHERE update_name = 'event_push_summary_unique_index';
```
Comment (DMRobertson) [inline synapse/storage/schema/main/delta/73/23_fix_thread_index.sql:25]: I sanity checked to see if we need to update `UNIQUE_INDEX_BACKGROUND_UPDATES`. It doesn't look like it: that only mentions the newest index, `event_push_summary_unique_index2`

**Review by DMRobertson — COMMENTED** [review]
Being fully paranoid: suppose that an old version of the source code is running the background update `event_push_summary_unique_index`. The admin upgrades the source code on disk and runs `upgrade_synapse_database` without stopping Synapse. (This means they must be running postgres.) Then we are concurrently running `CREATE INDEX CONCURRENTLY....` in one connection while running `DROP INDEX IF EXISTS...` in another. How do these interact?

(Even if the answer is "badly", I think this is unlikely to happen and we should probably ignore it?)

File: synapse/storage/schema/main/delta/73/23_fix_thread_index.sql (hunk 1 above)
Comment (clokep) [reply synapse/storage/schema/main/delta/73/23_fix_thread_index.sql:27]: Hah, thanks. I failed to remove it when finishing the comment. 👍

File: synapse/storage/schema/main/delta/73/23_fix_thread_index.sql (hunk 3 above)
Comment (clokep) [reply synapse/storage/schema/main/delta/73/23_fix_thread_index.sql:25]: Good idea. Thanks for checking!

File: synapse/storage/databases/main/event_push_actions.py (hunk 2 above)
Comment (clokep) [reply synapse/storage/databases/main/event_push_actions.py:285]: An error gets raised if it cannot find the background update, but as you've said we're deleting it so it doesn't matter.

**Comment by clokep** [comment]
> Being fully paranoid: suppose that an old version of the source code is running the background update `event_push_summary_unique_index`. The admin upgrades the source code on disk and runs `upgrade_synapse_database` without stopping Synapse. (This means they must be running postgres.) Then we are concurrently running `CREATE INDEX CONCURRENTLY....` in one connection while running `DROP INDEX IF EXISTS...` in another. How do these interact?
>
> (Even if the answer is "badly", I think this is unlikely to happen and we should probably ignore it?)

What I did was:

1. Create a table.
2. (Terminal A) Begin a transaction and lock the table exclusively.
3. (Terminal B) Start creating an index on that table (this hangs, waiting for the lock to clear).
4. (Terminal C) Drop the index, this gets skipped since the index doesn't exist.
5. (Terminal A) Commit the transaction.
6. (Terminal B) Note that the index was created.
7. Inspect the table and see that it *does* have the index.

So this interacts in such a way that it would leave us with the current problem.

**Review by DMRobertson — APPROVED** [review]
> What I did was:

It might be that a different serialisability level could avoid this? :shrug:

But as we said in #synapse-dev, let's get this in and hope it fixes up the majority of people who were bitten by this

**Pull request merged**